### PR TITLE
Catch stackoverflow errors in the highlighter

### LINKED ISSUE #16904: REPL gives stack overflow during too eager syntax highlighting, e.g. when multiplying string
## Compiler version

3.3.0-RC2

## Minimized code

```Scala
$ scala-cli repl -S 3.3.0-RC2
Welcome to Scala 3.3.0-RC2 (17.0.4.1, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.
                                                                                                                                                                       
scala> "works fine" * 100
val res0: String = works fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks fineworks ...
                                                                                                                                                                       
scala> "works not fine"* 1000
Exception in thread "main" java.lang.StackOverflowError
    at dotty.tools.dotc.ast.untpd$UntypedTreeAccumulator.foldMoreCases(untpd.scala:799)
    at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1641)
    at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
    at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
    at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
    at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
 

```

## Expectation

Should work and not give stack overflow (as in e.g. 2.13)

### LINKED ISSUE #18596: TreeHighlighter crash while generating an xml string at the console
I have tried `sbt console` as well as `scala-cli console`, in both cases, generating some valid xml strings crashes with a `StackOverflowError`. Running the same code without the console works as expected.

Also, generating a non-xml string just works (`(1 to 200).foldRight("x") { case (_, n) => s"< $n >" }`).

Previously, I was able to generate a different error which I can't reproduce again because I can't find which Scala version I used in that session:

error

## Compiler version

3.3.0

## Minimized example

```Scala
(1 to 200).foldRight("x") { case (_, n) => s"<x>$n</x>" }
```

## Output

```scala
Exception in thread "main" java.lang.StackOverflowError
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1545)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1545)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1545)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1557)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1545)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1551)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.apply(untpd.scala:808)
	at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1565)
	at dotty.tools.dotc.ast.untpd$UntypedTreeTraverser.traverseChildren(untpd.scala:809)
	at dotty.tools.dotc.printing.SyntaxHighlighting$TreeHighlighter$2$.traverse(SyntaxHighlighting.scala:123)
```

## Expectation

The console mustn't crash.

### PR DESCRIPTION
**Pull request by Valentin889**
# StackOverflow Exception on Highlight

Fixes #18596 and fixes #16904.

When there is too much instruction the Sytax Hyghlight can't make the tree and launch a StackOverflow Error

To fix the problem let's catch the Overflow and return the entry string with an error message

the lines that can triggers the error : 

```scala
val parser = new Parser(source)
val trees = parser.blockStatSeq()
TreeHighlighter.highlight(trees)
```

To fix it, let's put the try on everything starting from those lines and the catch in the end

```scala
case(e): StackOverflowError =>
          in
```

There is no obvious need of an error message, the command and result will simply not be highlighted so I choose to not put one there.

### CODE REVIEW

**Comment by mbovel** [comment]
There is a failure in SBT tests.

It seems to originate from Scaladoc:

```
Error:  Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.HashMap.get(Object)" is null
Error:  	at com.vladsch.flexmark.html.renderer.HeaderIdGenerator.generateId(HeaderIdGenerator.java:124)
Error:  	at com.vladsch.flexmark.html.renderer.HeaderIdGenerator.getId(HeaderIdGenerator.java:148)
Error:  	at dotty.tools.scaladoc.tasty.comments.markdown.SectionRenderingExtension$SectionHandler$.render(SectionRenderingExtension.scala:38)
Error:  	at dotty.tools.scaladoc.tasty.comments.markdown.SectionRenderingExtension$SectionHandler$.render(SectionRenderingExtension.scala:32)
Error:  	at com.vladsch.flexmark.html.renderer.NodeRenderingHandler.render(NodeRenderingHandler.java:16)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderNode(HtmlRenderer.java:779)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderChildrenNode(HtmlRenderer.java:798)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderChildren(HtmlRenderer.java:790)
Error:  	at com.vladsch.flexmark.html.renderer.CoreNodeRenderer.render(CoreNodeRenderer.java:113)
Error:  	at com.vladsch.flexmark.html.renderer.NodeRenderingHandler.render(NodeRenderingHandler.java:16)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderNode(HtmlRenderer.java:761)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.render(HtmlRenderer.java:683)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer.render(HtmlRenderer.java:213)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer.render(HtmlRenderer.java:197)
Error:  	at com.vladsch.flexmark.html.HtmlRenderer.render(HtmlRenderer.java:230)
Error:  	at dotty.tools.scaladoc.site.TemplateFile.resolveInner(templates.scala:128)
Error:  	at dotty.tools.scaladoc.site.LoadedTemplate.resolveToHtml(LoadedTemplate.scala:54)
Error:  	at dotty.tools.scaladoc.renderers.ResolvedTemplate.<init>(SiteRenderer.scala:18)
Error:  	at dotty.tools.scaladoc.renderers.ResolvedTemplate$.apply(SiteRenderer.scala:17)
```

Trying to re-run.

**Comment by hamzaremmal** [comment]
> There is a failure in SBT tests.
> 
> It seems to originate from Scaladoc:
> 
> ```
> Error:  Caused by: java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "java.util.HashMap.get(Object)" is null
> Error:  	at com.vladsch.flexmark.html.renderer.HeaderIdGenerator.generateId(HeaderIdGenerator.java:124)
> Error:  	at com.vladsch.flexmark.html.renderer.HeaderIdGenerator.getId(HeaderIdGenerator.java:148)
> Error:  	at dotty.tools.scaladoc.tasty.comments.markdown.SectionRenderingExtension$SectionHandler$.render(SectionRenderingExtension.scala:38)
> Error:  	at dotty.tools.scaladoc.tasty.comments.markdown.SectionRenderingExtension$SectionHandler$.render(SectionRenderingExtension.scala:32)
> Error:  	at com.vladsch.flexmark.html.renderer.NodeRenderingHandler.render(NodeRenderingHandler.java:16)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderNode(HtmlRenderer.java:779)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderChildrenNode(HtmlRenderer.java:798)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderChildren(HtmlRenderer.java:790)
> Error:  	at com.vladsch.flexmark.html.renderer.CoreNodeRenderer.render(CoreNodeRenderer.java:113)
> Error:  	at com.vladsch.flexmark.html.renderer.NodeRenderingHandler.render(NodeRenderingHandler.java:16)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.renderNode(HtmlRenderer.java:761)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer$MainNodeRenderer.render(HtmlRenderer.java:683)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer.render(HtmlRenderer.java:213)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer.render(HtmlRenderer.java:197)
> Error:  	at com.vladsch.flexmark.html.HtmlRenderer.render(HtmlRenderer.java:230)
> Error:  	at dotty.tools.scaladoc.site.TemplateFile.resolveInner(templates.scala:128)
> Error:  	at dotty.tools.scaladoc.site.LoadedTemplate.resolveToHtml(LoadedTemplate.scala:54)
> Error:  	at dotty.tools.scaladoc.renderers.ResolvedTemplate.<init>(SiteRenderer.scala:18)
> Error:  	at dotty.tools.scaladoc.renderers.ResolvedTemplate$.apply(SiteRenderer.scala:17)
> ```
> 
> Trying to re-run.

FYI: We have fixed this issue before the repo transfer (See #19884). You might consider rebasing this PR if you have this issue again (they were rare at the time)

**Comment by mbovel** [comment]
Thanks @hamzaremmal.

@Valentin889 can you please rebase this PR on `main` and see if it passes all tests?

**Comment by hamzaremmal** [comment]
> Thanks @hamzaremmal.
> 
> @Valentin889 can you please rebase this PR on `main` and see if it passes all tests?

@mbovel PR is already green (you started a second attempt). I don't think the rebasing is necessary.

**Comment by SethTisue** [comment]
what about https://github.com/scala/scala3/issues/19893 , could this be expanded a bit to improve the situation there too? it's a highlighting crash that isn't an SOE

**Comment by mbovel** [comment]
> what about #19893 , could this be broadened a bit to improve the situation there too? it's a highlighting crash that isn't an SOE

I don't know, catching stack overflows was already subject to debate, and catching more errors sound dangerous to me because we probably don't want to silence too many errors automatically. #19893 might be a bug that needs a separate fix, and we wouldn't have gotten a report for it if the error had been caught.

File: compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala [hunk 1]
```diff
@@ -124,24 +124,30 @@ object SyntaxHighlighting {
...
+          highlighted.append(in(idx))
+        }
+
+        if (colorAt.last != NoColor)
+          highlighted.append(NoColor)
 
-      highlighted.toString
+        highlighted.toString
+      catch
+        case(e): StackOverflowError =>
```
Comment (sjrd) [inline compiler/src/dotty/tools/dotc/printing/SyntaxHighlighting.scala:148]: This is a very strange style. We typically write such clauses like this:
```suggestion
        case e: StackOverflowError =>
```

**Pull request merged**